src/Spago/Cmd.purs

```diff
@@ -37,33 +91,38 @@ defaultExecOptions =
 
 spawn :: forall m. MonadAff m => String -> Array String -> ExecOptions -> m Execa.ExecaProcess
 spawn cmd args opts = liftAff do
-  subprocess <- Execa.execa cmd args (_ { cwd = opts.cwd })
+  let
+    stdinOpt = case opts.pipeStdin of
+      StdinPipeParent -> Just inherit
+      StdinWrite _ -> Just pipe
+      StdinNewPipe -> Just pipe
+  subprocess <- Execa.execa cmd args (_ { cwd = opts.cwd, stdin = stdinOpt, stdout = Just pipe, stderr = Just pipe })
 
   case opts.pipeStdin of
-    StdinPipeParent -> subprocess.stdin.shareParentProcessStdin
-    StdinWrite s -> subprocess.stdin.writeUtf8End s
+    StdinWrite s | Just { writeUtf8End } <- subprocess.stdin -> writeUtf8End s
     _ -> pure unit
+
   when (opts.pipeStderr) do
-    subprocess.stderr.pipeToParentStderr
+    traverse_ _.pipeToParentStderr subprocess.stderr
   when (opts.pipeStdout) do
-    subprocess.stdout.pipeToParentStdout
+    traverse_ _.pipeToParentStdout subprocess.stdout
 
   pure subprocess
 
-joinProcess :: forall m. MonadAff m => Execa.ExecaProcess -> m (Either ExecError ExecResult)
-joinProcess cp = liftAff $ cp.result
+joinProcess :: forall m. MonadAff m => Execa.ExecaProcess -> m ExecResult
```

Comment:
Why remove the `Either` here?

It's nice because it forces the consumer to unpack the result. Just having an `ExecResult` means that one can accidentally forget to check for `isSuccess`.

We could keep a `Either ExecResult ExecResult`, which would be functionally equivalent, but force consumers to unpack the result of the subprocess call.

Reply:
Performing an HTTP request is an operation that can succeed or fail (for a variety of reasons), and it makes sense that the result is encapsulated in an `Either`.
E.g. `Affjax` has `request :: forall a. AffjaxDriver -> Request a -> Aff (Either Error (Response a))`

We use languages like PureScript to have the typesystem - rather than human reviewers - catch things like "I forgot to unpack this result": humans get things wrong, the compiler doesn't risk forgetting.

> Needing to write either _.stdout _.stdout is annoying.

I think we can have a `Cmd.getStdout :: Either ExecResult ExecResult -> String` for cases like that one.

Reply:
Fair enough. I'll make the changes.